### DISCUSSION TITLE
Changing behavior of Auth / Guard Providers credential validation

### DISCUSSION THREAD

**Discussion by dievardump** [adonisjs/core#2110] [Help] [answered]
Hello,

I read the documentation and several issues here but there is no answer to my problem.

I wish to use the api guard with a custom OAT driver: instead of email/password authentication, users identify using signing message with private keys. They are identified in the DB by their public key. I need to go with this flow:

-> user request a message to sign (driver creates a unique message and associate it to user)
-> user send back message signed using their private key
-> server checks if recover(message, signature) === the public address the user claims to be 

To do this I need:
- to add an api point for the `message` request (that I should be able to manage) 
- override the way verifyCredentials works so the driver verifies that the message transmitted and the signature are the right one. I'm not sure how/where to override this. 

My understanding is that I need to create a Guard that extends OAT and redefines the validateCredential function (and maybe adds a createMessageToSign), but I'm not sure about this, and I don't understand how I can register it as a driver for auth.
 
Any guidance?

**Comment by dievardump** [comment]
So I've actually been able to do this.

First I created a provider that registers a new AuthManager which itself registers a new driver (and in auth.ts config I just change driver from `oat` to `signedmessage`)

```js
public register() {
    // Register AuthManager
    this.app.container.singleton('MyProvider/Auth', () => {
      const { AuthManager } = require('./src/AuthManager');

      return new AuthManager(this.app);
    });
  }

  public async boot() {
    this.app.container.with(
      ['Adonis/Addons/Auth', 'MyProvider/Auth'],
      (Auth, MyProviderAuth) => {
        Auth.extend('guard', 'signedmessage', function () {
          return MyProviderAuth.makeSignedMessageGuard(...arguments);
        });
      }
    );
  }
  ```
  
  Then makeSignedMessageGuard creates an instance of SignedMessageGuard
  
  ```js
  public makeSignedMessageGuard(_, mapping, config, provider, ctx) {
    const { SignedMessageGuard } = require('../Guards/SignedMessage');

    const tokenProvider = Auth.makeTokenProviderInstance(config.tokenProvider);
    const emitter = this.getEmitter();

    return new SignedMessageGuard(
      mapping,
      config,
      emitter,
      provider,
      ctx,
      tokenProvider,
      this.messageManager // this keeps message between calls
    ) as any;
  }
  ```
  
`SignedMessageGuard` simply extends `@adonisjs/auth/build/src/Guards/Oat`, defines `generateMessage` and redefines `verifyCredentials`

```js
  /**
   * Verifies user credentials
   */
  public async verifyCredentials(address: string, signature: string): Promise<any> {
    if (!address || !signature) {
      throw InvalidCredentialsException.invalidUid(this.name);
    }
    
    // verufy message is where message previously stored is tested against signature for current address
    await this.verifyMessage(address, signature);
    const providerUser = await this.lookupUsingUid(address);
    return providerUser.user;
  }
  
  // called preLogin to generate a message to sign for address
  public async generateMessage(address: string): string {
    return await this.messageManager.generateMessage(address);
  }
  ``` 
  
  This actually works as expected. Tokens are created.

**Reply by dievardump** [reply]
This is maybe not optimal, I'm not sure. But at least I can use OATGuard with something else than email/password which is already great.

Hope it helps anyone trying to do some custom Guard. You can just extend BaseGuard instead of OATGuard or any other. Pretty straightforward.

**Reply by dievardump** [reply]
However, as Virk pointed out on discord: 

> You won't have to override any methods
> Just do the verification in your controller and fetch a user instance
> And then call
> auth.login(user)

I should have just forget about using `auth.attempt` and just have done the verification against the message in my controller

**Comment by thetutlage** [comment, marked as answer]
I believe you went too far with this. Lets try to break down the authentication as a process.

- Verifying a user can be done in many ways. It can be `email + password`. 
- It can be `phone + otp` or any other way. AdonisJS provides a helper method `auth.attempt(email, password)` for the most common scenario.
- However, this is not the only way. You are free to verify your users as you like and once you have the `User instance` with you, you can call the following method to mark them as logged in 

```ts
await auth.login(user)
```

**Reply by dievardump** [reply]
Yes definitely, you are absolutely right, I just read your discord message and feel very dumb for not having thought about it.

I think the fact I didn't know I can just force a `auth.login(user)` didn't help. And When I saw it in the code I was already too far in the coding of a Guard to think about it

At least, now I have a base if I ever need to create a custom Guard, but I will sure change the way I do it and do the verification in the controller. thank you!
